README.md
# Nova Geolocation Field

This package for [Laravel Nova](https://nova.laravel.com/) provides an input control for selecting geo-coordinates on a map.
 In addition, it's possible to geocode addresses.

The field will not write values on the model. Instead, it will sync with fields for latitude and longitude you need to provide.

## Requirements

* `php: ^8.0`
* `laravel/nova: ^4.0`

## Installation / Getting started

Install the package via Composer:

```
composer require gabelbart/nova-geolocation-field
```

Publish static assets:
```
php  artisan vendor:publish --tag=laravel-assets --ansi
```

Set up your Resource class:

1. Add the `\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation` field
2. Add fields for latitude and longitude

```php
<?php

namespace App\Nova;

use Laravel\Nova\Fields\Hidden;
use Laravel\Nova\Fields\ID;
use Laravel\Nova\Http\Requests\NovaRequest;

use Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation;

class User extends Resource
{
    public static $model = \App\Models\User::class;
    public static $title = 'name';
    public static $search = ['id'];

    public function fields(NovaRequest $request): array
    {
        return [
            ID::make()->sortable(),

            Hidden::make('latitude'),
            Hidden::make('longitude'),
            Geolocation::make('Map position'),
        ];
    }

    // snip...
}
```

*Note: you can also use Number fields where the user can see/enter the selected latitude/longitude.
 The sync works in both directions.*

## Usage

### Options

#### `defaultLatitude(float $latitude)`

Set the default latitude for the map center.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->defaultLatitude(52.520008); // Berlin
```

#### `defaultLatitude(float $latitude)`

Set the default latitude for the map center.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->defaultLongitude(13.404954); // Berlin
```

#### `defaultZoom(int $zoom)`

Set the default [zoom](https://wiki.openstreetmap.org/wiki/Zoom_levels) for the map.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->defaultZoom(10);
```

#### `latitudeField(string $fieldName)`

Attribute-name of the latitude field.

```php
\Laravel\Nova\Fields\Number::make('Latitude', 'my_latitude');
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->latitudeField('my_latitude');
```

#### `longitudeField(string $fieldName)`

Attribute-name of the latitude field.

```php
\Laravel\Nova\Fields\Number::make('Longitude', 'my_longitude');
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->longitudeField('my_longitude');
```

#### `streetField(string $fieldName)`

Attribute-name of the street field.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->streetField('street');
```

#### `streetNumberField(string $fieldName)`

Attribute-name of the street-number/house-number field.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->streetNumberField('street_number');
```

#### `postalCodeField(string $fieldName)`

Attribute-name of the postal-code field.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->postalCodeField('postal_code');
```

#### `cityField(string $fieldName)`

Attribute-name of the city field.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->cityField('city');
```

#### `countryField(string $fieldName)`

Attribute-name of the country field.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->countryField('country');
```

#### `regionField(string $fieldName)`

Attribute-name of the region/state field.

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->regionField('region');
```

#### `selectionMode(string $fieldName)`

Change the selection mode.
 The position on the map can be selected either by moving the mal (`\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::SELECTION_MODE_MOVE`) 
 or by double-clicking the map (`\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::SELECTION_MODE_DBCLICK`).

*`SELECTION_MODE_MOVE` is the default setting.*

```php
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->selectionMode(\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::SELECTION_MODE_DBCLICK);
```

#### `enableGeocoding(bool $flag = true)`

Enable the geocoding feature.
 For the geocoding feature to work you need to install and set up `spatie/geocoder:^3`.
 Also, you need to provide at least the `streetField`, `cityField` and `countryField` option. 

```php
\String::make('Street');
\String::make('City');
\String::make('Country');
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->cityField('street')
  ->cityField('city')
  ->cityField('country')
  ->enableGeocoding();
```

#### `writeBackGeocodedAddress(bool $flag = true)`

This option requires configured address fields and enabled geocoding.
 When a geocoded address is selected, the values for the address components get written back to the input fields. 

```php
\String::make('Street');
\String::make('Street Number');
\String::make('City');
\String::make('Zip');
\String::make('Country');
\String::make('Region');
\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::make('Map position')
  ->writeBackGeocodedAddress()
  ->cityField('street')
  ->cityField('street_number')
  ->cityField('city')
  ->postalCodeField('zip')
  ->cityField('country')
  ->regionField('region')
  ->enableGeocoding();
```

### Change caching of geocoding results

By default, geocoding requests will be cached for 7 days.
 Each time the cache is hit again this period is refreshed.

#### Disable caching

Call the `\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::disableGeocodingCache` static method in your service provider to disable the cache.
 This will result in a higher load on the Google geocoding API thus probably costs will increase.

#### Change cache duration

To change the TTL of the geocoding cache call the `\Gabelbart\Laravel\Nova\Fields\Geolocation\Geolocation::geocodingCacheTtl` static method in your service provider.
 It accepts either a time in minutes or a valid `\DateInterval` instance. 


CHANGELOG.md
## 0.2.3 - 2023-05-05

**Fixes**
* Remove pointless debug log message

## 0.2.2 - 2023-03-29

**Fixes:**
* Fix issue when map was hidden on initialisation for example with [`eminiarts/nova-tabs`](https://github.com/eminiarts/nova-tabs)

# 0.2.1 - 2023-03-28

**Fixes:**
* Fix property type for `min-zoom` and `max-zoom`
* Fix default zoom not applied

## 0.2.0 - 2023-03-27

**Features:**
* Allow writing back components of the selected Address to the form

**Fixes:**
* Fix error when geocoding and selection is empty

## 0.1.0 - 2023-03-27

- Initial release


src/Geolocation.php
<?php

namespace Gabelbart\Laravel\Nova\Fields\Geolocation;

use Illuminate\Support\Facades\Log;

use Carbon\CarbonInterval;

use Laravel\Nova\Fields\Field;
use Laravel\Nova\Fields\SupportsDependentFields;
use Laravel\Nova\Http\Requests\NovaRequest;

class Geolocation extends Field
{
    use SupportsDependentFields;

    protected static bool $cacheGeocodingResults = true;
    protected static \DateInterval $geocodingCacheTtl;

    const DEFAULT_LATITUDE_FIELD = 'latitude';
    const DEFAULT_LONGITUDE_FIELD = 'longitude';
    const DEFAULT_LATITUDE = 52.520008;
    const DEFAULT_LONGITUDE = 13.404954;
    const DEFAULT_ZOOM = 10;
    const TILE_LAYER_OPENSTREETMAP = [
        'layer-type' => 'base',
        'name' => 'OpenStreetMap',
        'url' => 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
        'attribution' => '&copy; <a target="_blank" href="https://osm.org/copyright">OpenStreetMap</a> contributors'
    ];

    const SELECTION_MODE_DBCLICK = 'dbclick';
    const SELECTION_MODE_MOVE = 'move';

    public $component = 'geolocation-field';

    public function __construct($name, $attribute = null, callable $resolveCallback = null)
    {
        parent::__construct($name, $attribute, $resolveCallback);

        $this->latitudeField(static::DEFAULT_LATITUDE_FIELD);
        $this->longitudeField(static::DEFAULT_LONGITUDE_FIELD);

        $this->defaultLatitude(static::DEFAULT_LATITUDE);
        $this->defaultLongitude(static::DEFAULT_LONGITUDE);
        $this->defaultZoom(static::DEFAULT_ZOOM);
        $this->tileLayer(static::TILE_LAYER_OPENSTREETMAP);
    }

    public static function disableGeocodingCache(bool $flag = true)
    {
        static::$cacheGeocodingResults = !$flag;
    }
    public static function shouldCacheGeocodingResults(): bool
    {
        return static::$cacheGeocodingResults;
    }
    public static function geocodingCacheTtl(int|\DateInterval $duration)
    {
        static::$geocodingCacheTtl = is_int($duration)
            ? CarbonInterval::minutes($duration)
            : $duration;
    }
    public static function getGeocodingCacheTtl(): \DateInterval
    {
        return !empty(static::$geocodingCacheTtl)
            ? static::$geocodingCacheTtl
            : CarbonInterval::days(7);
    }

    public function defaultLatitude(float $latitude): static
    {
        return $this->withMeta([__FUNCTION__ => $latitude]);
    }

    public function defaultLongitude(float $longitude): static
    {
        return $this->withMeta([__FUNCTION__ => $longitude]);
    }

    public function defaultZoom(int $zoom): static
    {
        return $this->withMeta([__FUNCTION__ => $zoom]);
    }

    public function latitudeField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function longitudeField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function streetField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function streetNumberField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function postalCodeField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function cityField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function countryField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function regionField(string $fieldName): static
    {
        return $this->withMeta([__FUNCTION__ => $fieldName]);
    }

    public function tileLayer(array $config): static
    {
        return $this->withMeta([__FUNCTION__ => $config]);
    }

    public function selectionMode(string $config): static
    {
        return $this->withMeta([__FUNCTION__ => $config]);
    }

    public function writeBackGeocodedAddress(bool $flag = true): static
    {
        return $this->withMeta([__FUNCTION__ => $flag]);
    }

    public function writeBackGeocodedCountryLongFormat(bool $flag = true): static
    {
        return $this->withMeta([__FUNCTION__ => $flag]);
    }

    public function enableGeocoding(bool $flag = true): static
    {
        if (!class_exists('\Spatie\Geocoder\Facades\Geocoder')) {
            Log::warning('gabelbart/nova-geolocation-field:'
                . ' geocoding was enabled but the spatie/geocoder package is not installed!'
                . ' In order for geocoding to work you need to install it via `composer require spatie/geocoder:^3`!');
        }
        if (empty(config('geocoder.key'))) {
            Log::warning('gabelbart/nova-geolocation-field:'
                . ' geocoding was enabled but no api-key was set for spatie/geocoder!'
                . ' In order for geocoding to work you need to configure an api-key for Googles geocoding API'
                . ' via env variable `GOOGLE_MAPS_GEOCODING_API_KEY` or config `geocoder.key`!');
        }

        return $this->withMeta([__FUNCTION__ => $flag]);
    }

    /**
     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
     */
    protected function fillAttributeFromRequest(NovaRequest $request, $requestAttribute, $model, $attribute)
    {
    }

    /**
     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
     */
    public function resolve($resource, $attribute = null)
    {
        $latitudeField = $this->meta['latitudeField'];
        $longitudeField = $this->meta['longitudeField'];

        $latitude = $resource->{$latitudeField};
        $longitude = $resource->{$longitudeField};

        $this->value = json_encode([
            'latitude' => !is_null($latitude) ? (float) $latitude : null,
            'longitude' => !is_null($longitude) ? (float) $longitude : null,
        ]);
    }
}
